File: mkdocs/overrides/devbook/websockets/listen-transaction-flow.py
import asyncio
import json
import os

from websockets import connect

NODE_URL = os.getenv(
	'NODE_URL',
	'https://reference.symboltest.net:3001'
)
WS_URL = NODE_URL.replace('http', 'ws', 1) + '/ws'
print(f'Using node {NODE_URL}')

MONITOR_ADDRESS = os.getenv(
	'MONITOR_ADDRESS',
	'TCHBDENCLKEBILBPWP3JPB2XNY64OE7PYHHE32I'
)
print(f'Monitoring address: {MONITOR_ADDRESS}')


async def main():
	async with connect(WS_URL) as websocket:
		# Connect and receive uid
		response = json.loads(await websocket.recv())
		uid = response['uid']
		print(f'Connected to {WS_URL} with uid {uid}')

		# Subscribe to transaction channels
		channels = [
			f'confirmedAdded/{MONITOR_ADDRESS}',
			f'unconfirmedAdded/{MONITOR_ADDRESS}',
			f'unconfirmedRemoved/{MONITOR_ADDRESS}',
		]
		for channel in channels:
			await websocket.send(json.dumps(
				{'uid': uid, 'subscribe': channel}
			))
			name = channel.split('/')[0]
			print(f'Subscribed to {name} channel')

		# Handle incoming messages
		try:
			async for raw_message in websocket:
				message = json.loads(raw_message)
				topic = message['topic']
				transaction_hash = message['data']['meta']['hash']
				name = topic.split('/')[0]
				print(f'{name}: hash={transaction_hash[:16]}...')

		# Unsubscribe on exit
		finally:
			for channel in channels:
				await websocket.send(json.dumps({
					'uid': uid,
					'unsubscribe': channel
				}))
			print('Unsubscribed from all channels')


try:
	asyncio.run(main())
except KeyboardInterrupt:
	pass
except Exception as error:
	print(error)

File: mkdocs/overrides/devbook/websockets/listen-transaction-flow.mjs
const NODE_URL = process.env.NODE_URL
	|| 'https://reference.symboltest.net:3001';
const WS_URL = NODE_URL.replace('http', 'ws') + '/ws';
console.log(`Using node ${NODE_URL}`);

const MONITOR_ADDRESS = process.env.MONITOR_ADDRESS
	|| 'TCHBDENCLKEBILBPWP3JPB2XNY64OE7PYHHE32I';
console.log(`Monitoring address: ${MONITOR_ADDRESS}`);

const websocket = new WebSocket(WS_URL);

// Connect and receive uid
const uid = await new Promise((resolve) => {
	websocket.addEventListener('message', (event) => {
		const message = JSON.parse(event.data);
		resolve(message.uid);
	}, { once: true });
});
console.log(`Connected to ${WS_URL} with uid ${uid}`);

// Subscribe to transaction channels
const channels = [
	`confirmedAdded/${MONITOR_ADDRESS}`,
	`unconfirmedAdded/${MONITOR_ADDRESS}`,
	`unconfirmedRemoved/${MONITOR_ADDRESS}`,
];
for (const channel of channels) {
	websocket.send(JSON.stringify({ uid, subscribe: channel }));
	const name = channel.split('/')[0];
	console.log(`Subscribed to ${name} channel`);
}

// Handle incoming messages
websocket.addEventListener('message', (event) => {
	const message = JSON.parse(event.data);
	const topic = message.topic;
	const transactionHash = message.data.meta.hash;
	const name = topic.split('/')[0];
	console.log(`${name}: hash=${transactionHash.substring(0, 16)}...`);
});

// Unsubscribe on exit
process.on('SIGINT', () => {
	for (const channel of channels) {
		websocket.send(JSON.stringify({uid, unsubscribe: channel}));
	}
	console.log('Unsubscribed from all channels');
	websocket.close();
	process.exit(0);
});
